server/src/main/java/org/elasticsearch/search/aggregations/bucket/histogram/VariableWidthHistogramAggregator.java

```diff
@@ -52,25 +52,295 @@
+                    // The clusters in range {index ... numClusters - 1} move up 1 index to make room for the new cluster
+                    clusterMaxes.set(i, clusterMaxes.get(i-1));
+                    clusterMins.set(i, clusterMins.get(i-1));
+                    clusterCentroids.set(i, clusterCentroids.get(i-1));
+                    clusterSizes.set(i, clusterSizes.get(i-1));
+                }
+                clusterMaxes.set(idx, holdMax);
+                clusterMins.set(idx, holdMin);
+                clusterCentroids.set(idx, holdCentroid);
+                clusterSizes.set(idx, holdSize);
+
+                // Move the underlying buckets
+                long[] mergeMap = new long[numClusters];
+                for (int i = 0; i < idx; i++) {
+                    // The clusters in range {0 ... idx - 1} don't move
+                    mergeMap[i] = i;
+                }
+                for (int i = idx; i < numClusters - 1; i++) {
+                    // The clusters in range {index ... numClusters - 1} shift up
+                    mergeMap[i] = i + 1;
+                }
+                // Finally, the new cluster moves to index
+                mergeMap[numClusters - 1] = idx;
+
+                mergeBuckets(mergeMap, numClusters);
+                if (deferringCollector != null) {
+                    deferringCollector.mergeBuckets(mergeMap);
+                }
+            }
+        }
+
+        /**
+         * Add <code>val</code> to the cluster at index <code>bucketOrd</code>.
+         * The cluster's centroid, min, max, and size are recalculated.
+         */
+        private void addToCluster(int bucketOrd, double val){
+            assert bucketOrd < numClusters;
+
+            double max = Math.max(clusterMaxes.get(bucketOrd), val);
+            double min = Math.min(clusterMins.get(bucketOrd), val);
+
+            // Recalculate the centroid
+            double oldCentroid = clusterCentroids.get(bucketOrd);
+            double size = clusterSizes.get(bucketOrd);
+            double newCentroid = ((oldCentroid * size) + val) / (size + 1);
+
+            clusterMaxes.set(bucketOrd, max);
+            clusterMins.set(bucketOrd, min);
+            clusterCentroids.set(bucketOrd, newCentroid);
+            clusterSizes.increment(bucketOrd, 1);
+        }
+
+        /**
+         * Returns the ordinal of the bucket whose centroid is closest to <code>val</code>, or -1 if there are no buckets.
+         *
+         * This is done using a binary search, since <code>clusterCentroids</code> is always kept sorted.
+         * But, we don't care if <code>val</code> matches a bucket's key exactly. We just want the nearest bucket,
+         * and this requires some extra logic.
+         */
+        private int getNearestBucket(double val) {
+            if(numClusters == 0) return -1;
+
+            int left = 0;
+            int right = numClusters - 1;
+            while(left < right){
```

Comment:
I'm somewhat surprised we don't have a binary search implementation that properly targets `BigArrays`, but it looks like we don't have one. It *might* be better to just make one in the same vein as `org.apache.lucene.util.Sorter` just so you can write unit tests for *just* the binary search. I certainly don't trust myself to write a correct binary search without playing with a bunch of test cases. Simple though it may be there are sneaky overflow issues I never properly remember.